plugins/ExtendedTools/objmgr.c

```diff
@@ -420,393 +657,1548 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+            PhEndInitOnce(&initOnce);
+        }
+
+        if (NtOpenPartition_I)
+        {
+            status = NtOpenPartition_I(Handle, DesiredAccess, &objectAttributes);
+        }
+    }
+    else if (PhEqualString2(Context->Object->TypeName, L"CpuPartition", TRUE))
+    {
+        static PH_INITONCE initOnce = PH_INITONCE_INIT;
+        static NTSTATUS(NTAPI * NtOpenCpuPartition_I)(
+            _Out_ PHANDLE CpuPartitionHandle,
+            _In_ ACCESS_MASK DesiredAccess,
+            _In_ POBJECT_ATTRIBUTES ObjectAttributes
+            );
+
+        if (PhBeginInitOnce(&initOnce))
+        {
+            NtOpenCpuPartition_I = PhGetModuleProcAddress(L"ntdll.dll", "NtOpenCpuPartition");
+            PhEndInitOnce(&initOnce);
+        }
+
+        if (NtOpenCpuPartition_I)
+        {
+            status = NtOpenCpuPartition_I(Handle, DesiredAccess, &objectAttributes);
+        }
+    }
+    // Callback, Type (and almost any object type except ALPC Port and Device)
+    else
+    {
+        status = PhOpenObjectByTypeIndex(Handle, DesiredAccess, &objectAttributes, PhGetObjectTypeNumberZ(PhGetString(Context->Object->TypeName)));
+    }
+
+    NtClose(directoryHandle);
+
+    return status;
+}
+
+// Open real ALPC port by duplicating its "Connection" handle from the process that created the port
+// https://github.com/zodiacon/ObjectExplorer/blob/master/ObjExp/ObjectManager.cpp#L191
+// Open real FilterConnectionPort
+// Open real \REGISTRY key
+NTSTATUS EtpObjectManagerOpenRealObject(
+    _Out_ PHANDLE Handle,
+    _In_ PHANDLE_OPEN_CONTEXT Context,
+    _In_ ACCESS_MASK DesiredAccess,
+    _In_ PWSTR TypeName
+)
+{
+    static PH_INITONCE initOnce = PH_INITONCE_INIT;
+    static ULONG AlpcPortTypeIndex;
+    static ULONG FilterPortIndex;
+    static ULONG KeyTypeIndex;
+
+    if (PhBeginInitOnce(&initOnce))
+    {
+        AlpcPortTypeIndex = PhGetObjectTypeNumberZ(L"ALPC Port");
+        FilterPortIndex = PhGetObjectTypeNumberZ(L"FilterConnectionPort");
+        KeyTypeIndex = PhGetObjectTypeNumberZ(L"Key");
+        PhEndInitOnce(&initOnce);
+    }
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    PPH_STRING fullName;
+    ULONG TargetIndex = 0;
+
+    if (PhEqualStringZ(TypeName, L"ALPC Port", TRUE))
+        TargetIndex = AlpcPortTypeIndex;
+    else if (PhEqualStringZ(TypeName, L"FilterConnectionPort", TRUE))
+        TargetIndex = FilterPortIndex;
+    else if (PhEqualStringZ(TypeName, L"Key", TRUE))
+        TargetIndex = KeyTypeIndex;
+    else
+        return STATUS_INVALID_PARAMETER;
+
+    if (PhEqualStringRef(&Context->CurrentPath->sr, &EtObjectManagerRootDirectoryObject, TRUE))
+        fullName = PhFormatString(L"%s%s", PhGetString(Context->CurrentPath), PhGetString(Context->Object->Name));
+    else
+        fullName = PhFormatString(L"%s\\%s", PhGetString(Context->CurrentPath), PhGetString(Context->Object->Name));
+
+    PSYSTEM_HANDLE_INFORMATION_EX handles;
+
+    if (NT_SUCCESS(PhEnumHandlesEx(&handles)))
```

Comment:
This is both an expensive and invasive operation to try to duplicate a handle. I think we could find a more appropriate solution. Overall we're mostly dealing with "ALPC Port" and "FilterConnectionPort". I'm not sure why "Key" would require this path?

Maybe we should drop the ALPC and communication port parts until we come up with a nicer solution that isn't so invasive to the rest of the system?